### ISSUE TITLE
Improve performance on GTX 3080

### ISSUE THREAD

**Issue by proteneer** [proteneer/timemachine#343]
We're seeing a large (2x) performance drop on the GTX 3080 relative to a Titan V:

```
python tests/benchmark.py
```

For Apo the decrease is about 20%
```
hif2a-apo: N=8805 speed: 586.40ns/day dt: 1.5fs (ran 100000 steps in 22.11s)
hif2a-apo: N=8805 speed: 480.90ns/day dt: 1.5fs (ran 100000 steps in 26.95s)
```

For RBFE the decrease is a whopping 50%
```
Titan V: hif2a-rbfe: N=8840 speed: 471.76ns/day dt: 1.5fs (ran 100000 steps in 27.48s)
GTX 3080: hif2a-rbfe: N=8840 speed: 243.46ns/day dt: 1.5fs (ran 100000 steps in 53.24s)
```

Isolating the tests, running nsys nvprof on the two GPUs yields:

GTX 3080:
```
CUDA Kernel Statistics:

 Time(%)  Total Time (ns)  Instances    Average     Minimum    Maximum                                                   Name                                                
 -------  ---------------  ---------  -----------  ---------  ---------  ----------------------------------------------------------------------------------------------------
    78.6   56,862,044,216    103,000    552,058.7    529,310    569,951  void k_nonbonded_du_dx<float, true>(int, double const*, double const*, double const*, double, int c…
     3.1    2,205,410,088     17,380    126,893.6    106,143    158,880  void k_find_blocks_with_ixns<float>(int, double const*, double const*, double const*, double const*…
     2.3    1,633,281,666     86,900     18,795.0     17,856     19,936  void cub::RadixSortScanBinsKernel<cub::DeviceRadixSortPolicy<unsigned int, unsigned int, int>::Poli…
     1.9    1,380,093,966    104,010     13,268.9      8,352     15,744  void k_nonbonded_exclusions<float>(int, double const*, double const*, double const*, double, int co…
     1.7    1,219,826,475    103,000     11,843.0     10,880     12,576  void gen_sequenced<curandStateXORWOW, double2, normal_args_double_st, &(double2 curand_normal_scale…
     1.6    1,192,659,283      1,010  1,180,850.8  1,133,277  1,209,117  void k_nonbonded<float>(int, double const*, double const*, double const*, double, int const*, int c…                                                  
```

Titan V
```
Generating CUDA Kernel Statistics...
CUDA Kernel Statistics (nanoseconds)

Time(%)      Total Time   Instances         Average         Minimum         Maximum  Name                                                                                                                                                                                                                                                                                                                                         
-------  --------------  ----------  --------------  --------------  --------------  --------------------------------------------------------------------------------------------------------------------                                                                                                                                                                                                                         
   49.6     12203071831      102000        119638.0          112127          188415  void k_nonbonded_du_dx<float, true>(int, double const*, double const*, double const*, double, int const*, int const*, double, double, int const*, unsigned int const*, unsigned long long*)                                                                                                                                                  
    9.1      2244104571       17158        130790.6          107423          177311  void k_find_blocks_with_ixns<float>(int, double const*, double const*, double const*, double const*, unsigned int*, int*, unsigned int*, unsigned int*, double)                                                                                                                                                                              
    6.9      1687537239       85790         19670.6           18815           47712  void cub::RadixSortScanBinsKernel<cub::DeviceRadixSortPolicy<unsigned int, unsigned int, int>::Policy700, int>(int*, int)                                                                                                                                                                                                                    
    4.9      1212409938      103011         11769.7            6752           22623  void k_nonbonded_exclusions<float>(int, double const*, double const*, double const*, double, int const*, int const*, int const*, double const*, double, double, unsigned long long*, unsigned long long*, unsigned long long*, unsigned long long*)                                                                                          
    3.6       893742937       51474         17363.0           11904           30016  void cub::DeviceRadixSortDownsweepKernel<cub::DeviceRadixSortPolicy<unsigned int, unsigned int, int>::Policy700, true, false, unsigned int, unsigned int, int>(unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, int*, int, int, int, cub::GridEvenShare<int>)                                                          
```

**Comment by proteneer** [comment]
The nonbonded kernel in a 3080 is 5x slower than that of a Titan V! What the hell

**Comment by proteneer** [comment]
Isolating this down in the ampere_bench branch on a single kernel:

```
# Titan V
> nsys nvprof pytest -xvs tests/test_nonbonded.py::TestNonbondedDHFR::test_benchmark_3d
   80.6        19927131         100        199271.3          191614          216607  void k_nonbonded_du_dx<float, false>(int, double const*, double const*, double const*, double, int const*, int const*, double, double, int const*, unsigned int const*, unsigned long long*)                                                                                                                                                 
> nsys nvprof pytest -xvs tests/test_nonbonded.py::TestNonbondedDHFR::test_benchmark_4d
   83.1        20638174         100        206381.7          204543          209663  void k_nonbonded_du_dx<float, true>(int, double const*, double const*, double const*, double, int const*, int const*, double, double, int const*, unsigned int const*, unsigned long 

# GTX 3080
> nsys nvprof pytest -xvs tests/test_nonbonded.py::TestNonbondedDHFR::test_benchmark_3d
    91.8       57,306,938        100  573,069.4  570,175  576,031  void k_nonbonded_du_dx<float, false>(int, double const*, double const*, double const*, double, 
> nsys nvprof pytest -xvs tests/test_nonbonded.py::TestNonbondedDHFR::test_benchmark_4d
    95.7      102,949,107        100  1,029,491.1  1,026,462  1,034,141  void k_nonbonded_du_dx<float, true>(int, double const*, double const*, double const*, double, int c…
                                                                                                                                     
```

**Comment by proteneer** [comment]
Oh god they completely gutted the 64bit casting throughput, to a point where it's the worst it's ever been for *any* GPU ever released:


Type conversions from and to 64-bit types | 329 | 4 | 16 | 4 | 1610 | 16 | 2




 https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions

**Comment by proteneer** [comment]
This issue is mostly addressed by #347
